### ISSUE TITLE
Do we have any technical supplementary materials on NCCL code?

### ISSUE THREAD

**Issue by void-main** [NVIDIA/nccl#467]
Hi community, I'm trying to read the source code of NCCL, but found some code pretty hard to follow, for example, what's the meaning of `Search Patterns` in `ncclTopoCompute`, why we need `backToFirstRank`? I want to fully understand the reasons behind these code, do we have any docs / talks / blogs / references on the whole NCCL code?  Thank you!

**Comment by sjeaugey** [comment]
This part is by far the most complicated and documenting it would require to write a pretty long paper.

In short, the pattern is whether you want to find rings (closed loops), or chains (open). So backToFirstRank indicates we need to close the loop at the end to be considered a valid path since we're trying to find rings inside the node.

**Comment by void-main** [comment]
Thanks for your reply! Now I "feel like" I "kind of" understand the reason for `Search Patterns` according to your comments :-) Thanks!

Guess I'll have to dive deeper into the code without more reading materials. :-)
